lib/ansible/modules/crypto/openssl_privatekey_info.py

```diff
@@ -0,0 +1,667 @@
+        key_type = 'Ed25519'
+    elif CRYPTOGRAPHY_HAS_ED448 and isinstance(key, cryptography.hazmat.primitives.asymmetric.ed448.Ed448PrivateKey):
+        key_type = 'Ed448'
+    elif isinstance(key, cryptography.hazmat.primitives.asymmetric.ec.EllipticCurvePrivateKey):
+        key_type = 'ECC'
+        key_public_data['curve'] = key.public_key().curve.name
+        key_public_data['x'] = key.public_key().public_numbers().x
+        key_public_data['y'] = key.public_key().public_numbers().y
+        key_public_data['exponent_size'] = key.public_key().curve.key_size
+        key_private_data['multiplier'] = key.private_numbers().private_value
+    else:
+        key_type = 'unknown ({0})'.format(type(key))
+    return key_type, key_public_data, key_private_data
+
+
+def _binary_exp_mod(f, e, m):
+    """Computes f^e mod m."""
+    # Compute len_e = floor(log_2(e))
+    len_e = -1
+    x = e
+    while x > 0:
+        x >>= 1
+        len_e += 1
+    # Compute f**e mod m
+    result = 1
+    for k in range(len_e, -1, -1):
+        result = (result * result) % m
+        if ((e >> k) & 1) != 0:
+            result = (result * f) % m
+    return result
+
+
+def _simple_gcd(a, b):
+    '''Compute GCD of its two inputs.'''
+    while b != 0:
+        a, b = b, a % b
+    return a
+
+
+def _not_prime(n):
+    '''Does some quick checks to see if we can poke a hole into the primality of n.
+
+    A result of `False` does **not** mean that the number is prime; it just means
+    that we couldn't detect quickly whether it is not prime.
+    '''
+    if n <= 2:
+        return True
+    # The constant in the next line is the product of all primes < 200
+    if _simple_gcd(n, 7799922041683461553249199106329813876687996789903550945093032474868511536164700810) > 1:
+        return True
+    # TODO: maybe do some iterations of Miller-Rabin to increase confidence
+    # (https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test)
+    return False
+
+
+def _check_dsa_consistency(key_public_data, key_private_data):
+    # Get parameters
+    p = key_public_data.get('p')
+    q = key_public_data.get('q')
+    g = key_public_data.get('g')
+    y = key_public_data.get('y')
+    x = key_private_data.get('x')
+    for v in (p, q, g, y, x):
+        if v is None:
+            return None
+    # Make sure that g is not 0, 1 or -1 in Z/pZ
+    if g < 2 or g >= p - 1:
+        return False
+    # Make sure that x is in range
+    if x < 1 or x >= q:
+        return False
+    # Check whether q divides p-1
+    if (p - 1) % q != 0:
+        return False
+    # Check that g**q mod p == 1
+    if _binary_exp_mod(g, q, p) != 1:
+        return False
+    # Check whether g**x mod p == y
+    if _binary_exp_mod(g, x, p) != y:
+        return False
+    # Check (quickly) whether p or q are not primes
+    if _not_prime(q) or _not_prime(p):
+        return False
+    return True
+
+
+def _is_cryptography_key_consistent(key, key_public_data, key_private_data):
```

Comment:
Maybe these functions should go into `module_utils/crypto`? Seem useful for CSRs and certificates (maybe even SSH keys?) too.

Reply:
You mean the ones above this? Makes sense. I've added a commit for this; I've kept `_check_dsa_consistency` and `_is_cryptography_key_consistent`, though, since they are (at the moment) private key specific, and the latter specific to `cryptography`. We can still move them when we actually need them in other modules as well (and do some necessary generalizations then).

Reply:
Yes, the ones that verify properties of private/public keys.

I'm not 100% sure what would happen in the `selfsigned` and `ownca` providers if used with a keypair that violates these specs, but I agreee, that can be fixed later.